### PR TITLE
Bundle scala cli in scala command

### LINKED ISSUE #20098: Implement Scala CLI as the `scala` runner command
**Corresponding Scala CLI issue**
- https://github.com/VirtusLab/scala-cli/issues/2838

**Relevant docs**
- [SIP-46 details](https://github.com/scala/improvement-proposals/blob/main/content/scala-cli.md)
- [experimental Scala CLI as scala installation](https://scala-cli.virtuslab.org/docs/reference/scala-command/#testing-scala-cli-as-scala)
- [migration from the old Scala runner (a guide)](https://scala-cli.virtuslab.org/docs/reference/scala-command/#migrating-from-the-old-scala-runner-to-scala-cli)

**Requirements**
- [x] set up downloading of Scala CLI JVM launcher from GitHub VirtusLab/scala-cli
- [x] download library dependencies to the dist/target/pack in maven repo style
- [x] fix scripts to wrap  Scala CLI launcher, setting both the local repo and default scala version launcher options
  - [x] update for windows (batch)
  - [x] update for linux/mac (bash)
- [ ] run any existing legacy (old) `scala` runner tests against the Scala CLI launcher on the `scala/scala3` CI
- [ ] use the Scala CLI launcher instead of the `scala` runner in official Scala 3 distributions
  - [x] coursier
  - [ ] brew
  - [ ] SDKMAN

### PR DESCRIPTION
**Pull request by bishabosha**
fixes #20098

Proposed changes to zip/targz archive:
- in the `/bin` directory store an extra launcher for Scala CLI (either JAR, or native per platform).
- `/bin/scala[.bat]` is modified to invoke Scala CLI stored in `/bin`
- new `/maven2` directory, which stores all the Jars and POM files necessary (in maven repo style) for scala-cli to invoke scala compiler offline (using the `-r` launcher option).
- CHOICE: either replace jar files in `/lib` by aliases to the corresponding jar in `/maven2`, OR delete `/lib` and update references from scripts. (Looks like symlinks are not portable, so probably we should encode the classpath in a file, or adjust slightly how we build the toolchain)
- add platform specific suffixes to artefacts:
  - e.g. `scala-3.5.0-x86_64-pc-linux.tar.gz` (for the artefact that bundles the x64 linux launcher)

### CODE REVIEW

File: compiler/test-resources/scripting/showArgs.sc [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env scala
+#!/usr/bin/env bin/scala
 
 // precise output format expected by BashScriptsTests.scala
 def main(args: Array[String]): Unit =
```
Comment (hamzaremmal) [inline compiler/test-resources/scripting/showArgs.sc:4]: Why not here ?
```suggestion
// def main(args: Array[String]): Unit = // MIGRATION: Scala CLI expects `*.sc` files to be straight-line code
```

Comment (hamzaremmal) [reply compiler/test-resources/scripting/showArgs.sc:4]: Oh, but if we comment this then it will be the following test compiler/test-resources/scripting/showArgsNu.sc, what is the purpose of this test then ?

File: bin/scala [hunk 2]
```diff
@@ -2,4 +2,4 @@
 
 ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")" >& /dev/null && pwd)/.."
 
-"$ROOT/bin/common" "$ROOT/dist/target/pack/bin/scala" "$@"
+"$ROOT/bin/common" "$ROOT/dist/target/pack/bin/scala" "--power" "$@" "--offline" "--server=false"
```
Comment (sjrd) [inline bin/scala:5]: Put the forwarded arguments last?
```suggestion
"$ROOT/bin/common" "$ROOT/dist/target/pack/bin/scala" "--power" "--offline" "--server=false" "$@"
```

File: compiler/test/dotty/tools/scripting/ClasspathTests.scala [hunk 3]
```diff
@@ -51,7 +51,7 @@ class ClasspathTests:
       // convert scriptCp to a list of files
       val hashbangJars: List[File] = scriptCp.split(psep).map { _.toFile }.toList
       val hashbangClasspathJars = hashbangJars.map { _.name }.sorted.distinct // get jar basenames, remove duplicates
-      val packlibDir = s"$scriptCwd/$packLibDir" // classpathReport.sc specifies a wildcard classpath in this directory
+      val packlibDir: String = ??? /* ??? was s"$scriptCwd/$packLibDir" */ // classpathReport.sc specifies a wildcard classpath in this directory
```
Comment (sjrd) [inline compiler/test/dotty/tools/scripting/ClasspathTests.scala:54]: Erm ... how does this not crash?

File: compiler/test-resources/scripting/showArgs.sc (hunk 1 above)
Comment (bishabosha) [reply compiler/test-resources/scripting/showArgs.sc:4]: all files in /scripting are ran by CoursierScalaTests, which are run by the old ScriptingDriver so require a main method

**Review by sjrd — COMMENTED** [review]
I can only comment on what's jumping at me. The timing is way too short to do a real review.

Even if I did find something it wouldn't get addressed and shipped in time at this point, so, to be blunt: why bother?

File: bin/scala (hunk 2 above)
Comment (bishabosha) [reply bin/scala:5]: we do it this way because sub commands such as `run`, `package` (from `$@`) have to come first

File: compiler/test/dotty/tools/scripting/ClasspathTests.scala (hunk 3 above)
Comment (bishabosha) [reply compiler/test/dotty/tools/scripting/ClasspathTests.scala:54]: this is in an ignored test, but I deleted that method, it wasn't worth rewriting an ignored test

**Review by hamzaremmal — APPROVED** [review]

File: compiler/test-resources/scripting/showArgs.sc (hunk 1 above)
Comment (bishabosha) [reply compiler/test-resources/scripting/showArgs.sc:4]: I guess we may want to rethink those coursier tests once it can download the new launcher

**Comment by bishabosha** [comment]
It appears that on windows a critical class is loaded on boot `scala.build.internal.JniGetWinDirs` which is compiled by JDK 16. But on Mac OS the jar launcher works fine - could it be possible that just this one dependency is misconfigured?

**Comment by He-Pin** [comment]
A question is if a new scala-cli is released, will scala update it automatically, like `cs update scala-cli`?

**Comment by bishabosha** [comment]
> A question is if a new scala-cli is released, will scala update it automatically, like cs update scala-cli?

The way it is currently implemented, the idea would be that each scala version is tied to a specific Scala CLI launcher.

**Comment by He-Pin** [comment]
1. When do republishing , is there some kind of Sha or md5 checking ?
2. Does Windows platform been well tested?

**Comment by hamzaremmal** [comment]
- [ ] [README.md](https://github.com/scala/scala3/blob/7d559ad06bdfe448450341c7fdf12cc0513456f3/README.md) needs to be updated too

File: project/DottyVersion.scala [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package versionhelpers
+
+object DottyVersion {
+  val isRelease = sys.env.get("RELEASEBUILD") == Some("yes")
+
+  val baseVersion = "3.5.0-RC1"
+
+  val referenceVersion = "3.4.2-RC1"
```
Comment (He-Pin) [inline project/DottyVersion.scala:8]: Is it right?

Comment (bishabosha) [reply project/DottyVersion.scala:8]: we need a rebase

**Comment by bishabosha** [comment]
@hamzaremmal this is everything that was agreed in the plan, except for deciding what to do about the lib directory - and maybe caching the task to populate the local maven2 repo with coursier

**Comment by hamzaremmal** [comment]
> We would also need to update the ci.yml to publish artefacts for each launcher - and the sdkman publishing, but perhaps that can be another PR?

Yes, it should be in another PR with the changes to the launchers.yml workflow too. I will take care of it.

**Comment by bishabosha** [comment]
oh right and we should squash when merging this one I think

**Comment by hamzaremmal** [comment]
> oh right and we should squash when merging this one I think

Definitely

**Comment by bishabosha** [comment]
I have manually squashed the commits where it makes sense

File: project/scripts/winCmdTests [hunk 5]
```diff
@@ -1,10 +1,10 @@
...
 set -e
-PREFIX="dist/target/pack"
+PREFIX="dist/win-x86_64/target/pack"
 SOURCE="tests/pos/HelloWorld.scala"
 $PREFIX/bin/scalac @project/scripts/options "$SOURCE"
 $PREFIX/bin/scalac -d out "$SOURCE"
-$PREFIX/bin/scala -classpath out HelloWorld
-$PREFIX/bin/scala -classpath out -J-Xmx512m HelloWorld
+$PREFIX/bin/scala --power -classpath out -M HelloWorld --offline '--server=false'
+$PREFIX/bin/scala --power -classpath out -J -Xmx512m -M HelloWorld --offline '--server=false'
```
Comment (hamzaremmal) [inline project/scripts/winCmdTests:9]: Same comment as of the `project/scripts/native-integration/winTests.bat`. Why not clearing `scala-cli` folders and files ?

File: project/scripts/winCmdTests [hunk 6]
```diff
@@ -1,10 +1,10 @@
 #!/usr/bin/env bash
 
 set -e
-PREFIX="dist/target/pack"
+PREFIX="dist/win-x86_64/target/pack"
```
Comment (hamzaremmal) [inline project/scripts/winCmdTests:4]: Windows ? Shoudn't it be the linux plateform (this is a bash script) ?

File: project/scripts/winCmdTests.bat [hunk 7]
```diff
@@ -2,7 +2,7 @@
 setlocal
 
 @rem paths are relative to the root project directory
-set "_PREFIX=dist\target\pack"
+set "_PREFIX=dist\win-x86_64\target\pack"
```
Comment (hamzaremmal) [inline project/scripts/winCmdTests.bat:5]: I understand that the native version is tested here `project/scripts/native-integration/winTests.bat`. We should probably revert this change and keep testing with the `jar` version

File: project/scripts/winCmdTests.bat [hunk 8]
```diff
@@ -14,10 +14,10 @@ if not %ERRORLEVEL%==0 endlocal& exit /b 1
 call "%_PREFIX%\bin\scalac.bat" -d "%_OUT_DIR%" "%_SOURCE%"
 if not %ERRORLEVEL%==0 endlocal& exit /b 1
 
-call "%_PREFIX%\bin\scala.bat" -classpath "%_OUT_DIR%" HelloWorld
+call "%_PREFIX%\bin\scala.bat" --power -classpath "%_OUT_DIR%" -M HelloWorld --offline --server=false
 if not %ERRORLEVEL%==0 endlocal& exit /b 1
 
-call "%_PREFIX%\bin\scala.bat" -classpath "%_OUT_DIR%" -J-Xmx512m HelloWorld
+call "%_PREFIX%\bin\scala.bat" --power -classpath "%_OUT_DIR%" -J -Xmx512m -M HelloWorld --offline --server=false
 if not %ERRORLEVEL%==0 endlocal& exit /b 1
```
Comment (hamzaremmal) [inline project/scripts/winCmdTests.bat:21]: Clearing `scala-cli` files and folders too here

**Review by hamzaremmal — CHANGES_REQUESTED** [review]

File: bin/scala [hunk 9]
```diff
@@ -2,4 +2,37 @@
...
+        shift
+        ;;
+    esac
+  done
+
+  if [ $DISABLE_BLOOP -eq 1 ]; then
+    CLI_ARGS+=("--offline" "--server=false")
+  fi
+
+  echo "--power ${CLI_ARGS[@]} ${SCRIPT_ARGS[@]}"
```
[the remark this answers is not in the document]
Comment (bishabosha) [reply bin/scala:35]: echo is how you return the value from the function

File: compiler/test/dotty/tools/io/ClasspathTest.scala [hunk 10]
```diff
@@ -23,7 +25,7 @@ class ClasspathTest {
   @Test def testWildcards(): Unit =
     val outDir = Files.createTempDirectory("classpath-test")
     try
-      val compilerLib = "dist/target/pack/lib"
+      val compilerLib = s"${if isWindows then "dist-win-x86_64" else "dist"}/target/pack/lib"
```
[the remark this answers is not in the document]
Comment (bishabosha) [reply compiler/test/dotty/tools/io/ClasspathTest.scala:28]: because windows runners don't work with the jvm - but it should probably be the native launcher for each platform

File: project/scripts/winCmdTests.bat (hunk 7 above)
Comment (bishabosha) [reply project/scripts/winCmdTests.bat:5]: but the whole reason we do the native runners is because the windows runner only has jvm 8 on it so the test breaks

File: project/scripts/native-integration/winTests.bat [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+@rem if-tests mimic the non-existing bash instruction 'set -e'.
+call "%_PREFIX%\bin\scalac.bat" "@project\scripts\options" "%_SOURCE%"
+if not %ERRORLEVEL%==0 endlocal& exit /b 1
+
+call "%_PREFIX%\bin\scalac.bat" -d "%_OUT_DIR%" "%_SOURCE%"
+if not %ERRORLEVEL%==0 endlocal& exit /b 1
+
+call "%_PREFIX%\bin\scala.bat" --power -classpath "%_OUT_DIR%" -M HelloWorld --offline --server=false
+if not %ERRORLEVEL%==0 endlocal& exit /b 1
+
```
[the remark this answers is not in the document]
Comment (bishabosha) [reply project/scripts/native-integration/winTests.bat:18]: if you just run a class then it shouldn't create anything

File: project/scripts/winCmdTests (hunk 6 above)
Comment (bishabosha) [reply project/scripts/winCmdTests:4]: this is a windows test that is run on windows, it uses the bash that comes with Cygwin and the msys bash that comes with Git

File: project/scripts/winCmdTests.bat (hunk 8 above)
Comment (bishabosha) [reply project/scripts/winCmdTests.bat:21]: same thing here, we are only running a class, so no files are created

File: bin/scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (bishabosha) [reply bin/scala:35]: this is just a test command used only for testing, it automatically adds `--power`, `--offline` and `--server=false` to commands that support it. this isn't shipped with code

File: compiler/test-coursier/dotty/tools/coursier/CoursierScalaTests.scala [hunk 12]
```diff
@@ -166,7 +166,7 @@ object CoursierScalaTests:
       case Nil => args
       case _ => "--" +: args
     val newJOpts = jOpts.map(s => s"--java-opt ${s.stripPrefix("-J")}").mkString(" ")
-    execCmd("./cs", (s"""launch "org.scala-lang:scala3-compiler_3:${sys.env("DOTTY_BOOTSTRAPPED_VERSION")}" $newJOpts --main-class "$entry" --property "scala.usejavacp=true"""" +: newOptions)*)._2
+    execCmd("./cs", (s"""launch "org.scala-lang:scala3-compiler_3:${sys.env("DOTTY_BOOTSTRAPPED_VERSION")}" $newJOpts --main-class "$entry" --property "scala.usejavacp=true" --property "scala.use_legacy_launcher=true"""" +: newOptions)*)._2
```
[the remark this answers is not in the document]
Comment (bishabosha) [reply compiler/test-coursier/dotty/tools/coursier/CoursierScalaTests.scala:169]: cs launch always runs the main class, it doesn't download the GitHub launcher

File: tests/cmdTest-sbt-tests/sourcepath-with-inline/src/main/scala/a/zz.scala
[the remark this answers is not in the document]
Comment (bishabosha) [reply tests/cmdTest-sbt-tests/sourcepath-with-inline/src/main/scala/a/zz.scala:1]: im not sure how it got there, it might be an artefact, but it is used in bootstrappedCmdTests

File: project/scripts/cmdTestsCommon.inc.sh [hunk 13]
```diff
@@ -9,11 +9,15 @@ SOURCE="tests/pos/HelloWorld.scala"
 MAIN="HelloWorld"
 TASTY="HelloWorld.tasty"
 EXPECTED_OUTPUT="hello world"
+EXPECTED_OUTPUT_ARGS="[0:abc],[1:true],[2:123]"
```
[the remark this answers is not in the document]
Comment (bishabosha) [reply project/scripts/cmdTestsCommon.inc.sh:12]: its testing passing arguments to a script

File: project/scripts/echoArgs.sc
[the remark this answers is not in the document]
Comment (bishabosha) [reply project/scripts/echoArgs.sc:1]: yes its used there

File: project/scripts/native-integration/bashTests
[the remark this answers is not in the document]
Comment (bishabosha) [reply project/scripts/native-integration/bashTests:1]: do a text search for `bashTests`, you will see it is used in launchers.yml

File: project/scripts/native-integration/bashTests [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+test "$SCALA_VERSION" = "$std_output"
+
+echo "testing run command"
+std_output=$("$PROG_HOME/bin/scala" run "$SOURCE" --power --offline --server=false)
+test "hello world" = "$std_output"
+clear_cli_dotfiles "$ROOT/tests/pos"
+
+echo "testing run command (-with-compiler)"
+std_output=$("$PROG_HOME/bin/scala" run "$SOURCE_VERSION" -with-compiler --power --offline --server=false)
+test "$SCALA_VERSION" = "$std_output"
```
[the remark this answers is not in the document]
Comment (bishabosha) [reply project/scripts/native-integration/bashTests:82]: its clearly used in launchers.yml

File: project/scripts/buildScalaBinary [hunk 15]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+
+set -e
+
+source $(dirname $0)/cmdTestsCommon.inc.sh
```
[the remark this answers is not in the document]
Comment (bishabosha) [reply project/scripts/buildScalaBinary:5]: the $ROOT variable is not defined?

File: bin/scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (hamzaremmal) [reply bin/scala:35]: Yep, this is actually a stale comment that I forgot to remove. Sorry for that

File: compiler/test/dotty/tools/io/ClasspathTest.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (hamzaremmal) [reply compiler/test/dotty/tools/io/ClasspathTest.scala:28]: We should probably have tests for the jvm launcher and tests for all native launchers

File: project/scripts/native-integration/winTests.bat (hunk 11 above)
[the remark this answers is not in the document]
Comment (hamzaremmal) [reply project/scripts/native-integration/winTests.bat:18]: Right. `scala-cli` indeed doesn't generate files when running a main class

File: project/scripts/winCmdTests (hunk 6 above)
Comment (hamzaremmal) [reply project/scripts/winCmdTests:4]: Indeed: https://github.com/scala/scala3/blob/04ada792f2cc81177c2f430e1a1b9060eab7b54c/.github/workflows/ci.yaml#L237

File: project/scripts/native-integration/bashTests
[the remark this answers is not in the document]
Comment (hamzaremmal) [reply project/scripts/native-integration/bashTests:1]: I assumed that `launchers.yml` was unchanged since I've added it

File: project/scripts/echoArgs.sc
[the remark this answers is not in the document]
Comment (hamzaremmal) [reply project/scripts/echoArgs.sc:1]: Indeed, line 108

File: project/scripts/native-integration/bashTests (hunk 14 above)
[the remark this answers is not in the document]
Comment (hamzaremmal) [reply project/scripts/native-integration/bashTests:82]: Okay, but `launchers.yml` was not run on this PR and I don't see why `scala run $SOURCE_VERSION` would work. I doubt that there is a file with the same name as $SOURCE_VERSION

File: project/scripts/buildScalaBinary (hunk 15 above)
[the remark this answers is not in the document]
Comment (hamzaremmal) [reply project/scripts/buildScalaBinary:5]: I will change it by this:

```scala
ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")" >& /dev/null && pwd)/../.."

# set the $DIST_PROJECT and $DIST_DIR variables
source "$ROOT/bin/common-platform"
```

[the remark this answers is not in the document]
Comment (hamzaremmal) [reply project/scripts/buildScalaBinary]: For the CI, we also have to add this: 
```bash
SBT="$ROOT/project/scripts/sbt" # if run on CI
```

File: project/scripts/native-integration/bashTests (hunk 14 above)
[the remark this answers is not in the document]
Comment (hamzaremmal) [reply project/scripts/native-integration/bashTests:82]: I've triggered the `launchers.yml` workflow on this PR, lets see how it goes

[the remark this answers is not in the document]
Comment (hamzaremmal) [reply project/scripts/native-integration/bashTests:82]: Nevermind, I misunderstood what `$SOURCE_VERSION` was. I've also enabled the `launchers.yml` workflow on this PR. It still needs to be configured to use our runners instead of GitHub's

File: compiler/test/dotty/tools/io/ClasspathTest.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (bishabosha) [reply compiler/test/dotty/tools/io/ClasspathTest.scala:28]: but the jvm tests crash on windows because its JDK 8

[the remark this answers is not in the document]
Comment (hamzaremmal) [reply compiler/test/dotty/tools/io/ClasspathTest.scala:28]: What I meant is to these tests with each native-launcher on each platform and then run the jvm-launcher on linux. But maybe it is a bit overkill to do that at the moment

File: bin/scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (Gedochao) [reply bin/scala:35]: I still don't feel this is necessary... passing those arguments explicitly in the tests would make them more readable IMO, rather than embedding them in this script.
it's also easy to get confused where this script is being used.
This is a nitpick, however, we don't need to refactor it in this PR I think.

File: bin/test/TestScripts.scala [hunk 16]
```diff
@@ -57,7 +57,7 @@ class TestScripts {
       s"bin/scalac script did not run properly. Output:$lineSep$dotcOutput"
     )
 
-    val (retDotr, dotrOutput) = executeScript("./bin/scala HelloWorld")
+    val (retDotr, dotrOutput) = executeScript("./bin/scala -M HelloWorld")
```
Comment (Gedochao) [inline bin/test/TestScripts.scala:60]: this is slightly confusing, since just passing `-M` to Scala CLI, without any inputs or `-cp` will just fail.
I know it doesn't fail because `-cp` is passed elsewhere, but this line looks like it should fail, technically.

File: compiler/src/dotty/tools/MainGenericRunner.scala [hunk 17]
```diff
@@ -266,6 +266,22 @@ object MainGenericRunner {
           run(settings.withExecuteMode(ExecuteMode.Run))
         else
           run(settings.withExecuteMode(ExecuteMode.Repl))
+    end run
+
+    val ranByCoursierBootstrap =
+      sys.props.isDefinedAt("coursier.mainJar")
+      || sys.props.get("bootstrap.mainClass").contains("dotty.tools.MainGenericRunner")
+
+    val silenced = sys.props.get("scala.use_legacy_launcher") == Some("true")
```
Comment (Gedochao) [inline compiler/src/dotty/tools/MainGenericRunner.scala:275]: We may want to document this prop clearly somewhere later, once we release this.
I'm also thinking an env variable could potentially be useful?
`LEGACY_SCALA_RUNNER=true` or smth.

File: compiler/test-resources/scripting/classpathReport_scalacli.sc [hunk 18]
```diff
@@ -1,8 +1,8 @@
-#!bin/scala -classpath 'dist/target/pack/lib/*'
-
+#!/usr/bin/env bin/scala
+// This file is a Scala CLI script.
 import java.nio.file.Paths
 
-def main(args: Array[String]): Unit =
+// def main(args: Array[String]): Unit = // MIGRATION: Scala CLI expects `*.sc` files to be straight-line code
```
Comment (Gedochao) [inline compiler/test-resources/scripting/classpathReport_scalacli.sc:5]: I wonder if it makes sense to keep this comment, or can we just remove it.
it's stated earlier that this is a Scala CLI script.
```suggestion
```

File: compiler/test/dotty/tools/io/ClasspathTest.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (Gedochao) [reply compiler/test/dotty/tools/io/ClasspathTest.scala:28]: Do we still care about testing the JVM launcher on Windows?

**Review by Gedochao — APPROVED** [review]
LGTM
I just commented on some nitpicks, which can be addressed later in a future PR.

**Comment by hamzaremmal** [comment]
> Do we still care about testing the JVM launcher on Windows?

(Not sure why I can't reply directly to your comment) We can't run JVM launchers on Windows due to the restriction on the Java version. We will keep releasing the JVM launcher anyways for other uses so it is good to test them too.
